### ISSUE TITLE
Center Loss function, is it correct?

### ISSUE THREAD

**Issue by offbit** [davidsandberg/facenet#65]
https://github.com/davidsandberg/facenet/blob/master/src/facenet.py#L75

Please correct me if I'm wrong, but the paper creates a centre for each class in the training. The purpose is to gather each of class observation around the same area. In this loss function implementation it looks like 1 centre is created. So basically it forces all classes to be close to a single point. 

Did I understood this correctly ? 

Edit:
there is a stack-overflow post about this
http://stackoverflow.com/questions/40162361/how-to-implement-center-loss-and-other-running-averages-of-labeled-embeddings

**Comment by davidsandberg** [comment]
The dimension of the tensor `centers` in the center_loss function should have the dimensionality equal to the size of the logits, which is the second dimension of the `logits` tensor (since the first one is batch size).
When I run the `facenet_train_classifier.py`with center_loss>0.0 I indeed get a tensor with dimension 1792 (logits size for the inception_resnet_v1). 
On the wiki page [Classifier-training-of-inception-resnet-v1](https://github.com/davidsandberg/facenet/wiki/Classifier-training-of-inception-resnet-v1#7-plotting-learning-curves) the learning curves (including ones with center loss) is plotted and it can be seen that non-zero center loss gives a significant boost in performance.

**Comment by offbit** [comment]
I'm just comparing to the suggested solution by the paper.
https://github.com/ydwen/caffe-face/blob/caffe-face/src/caffe/layers/center_loss_layer.cpp#L64
which performs X(i,:) - C(y(i),:) (which is the distance to centre of the class in the embedding space). 
I've tried the loss function in a setup of my own, and with very low coef for center loss yes it should work, because practically it keeps embeddings around a central point. I'm not arguing about the result. But from the code point of view, it looks different. 

Thanks for the reply.

**Comment by offbit** [comment]
@davidsandberg  I can share with you the snippets of my implementation. Probably this weekend.

**Comment by davidsandberg** [comment]
@offbit Any news on this? I guess the numpy-style broadcasting in tensorflow makes the code look a bit different.

**Comment by offbit** [comment]
@davidsandberg 
Shall I paste the code here? 

declare on cpu this
`        with tf.variable_scope('embed'):
            embed_ctrs = tf.get_variable("ctrs", [dataop.NUM_LABELS, 512], dtype=tf.float32,
                                         initializer=tf.constant_initializer(0), trainable=False)
`

center loss function is this:
`
def center_loss(logits, labels):
    """Center loss based on the paper "A Discriminative Feature Learning Approach for Deep Face Recognition"
    (http://ydwen.github.io/papers/WenECCV16.pdf)
    """
    with tf.variable_scope('embed', reuse=True):
        centers = tf.get_variable("ctrs")

    label_batch = tf.reshape(labels, [-1])
    embed_ctrs_batch = tf.gather(centers, label_batch)
    loss = tf.nn.l2_loss(embed_ctrs_batch - logits)
    loss = tf.reduce_mean(loss)
    return loss
`
and you need a center update function after each forward/backward pass
`
def get_embed_centers(embed_batch, label_batch):
    """ Exponential moving window average. Increase decay for longer windows [0.0 1.0]"""
    decay = 0.97
    with tf.variable_scope('embed', reuse=True):
        embed_ctrs = tf.get_variable("ctrs")

    label_batch = tf.reshape(label_batch, [-1])
    old_embed_ctrs_batch = tf.gather(embed_ctrs, label_batch)
    dif = (1 - decay) * (old_embed_ctrs_batch - embed_batch)
    embed_ctrs = tf.scatter_sub(embed_ctrs, label_batch, dif)
    embed_ctrs_batch = tf.gather(embed_ctrs, label_batch)
    return embed_ctrs_batch
`

the center update is a bit different than the proposed in the center in the paper in terms of the weighted step. 

This was my implementation. So basically every class has it's own center that it's computed by the mini-batches. 
the loss is computed by the distance to center for each class. 

I hope it helped! 
I didn't have the time to thoroughly test it on faces. I'm doing a different application.
